core/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java

```diff
@@ -0,0 +1,209 @@
+ */
+public abstract class TransportWriteAction<
+            Request extends ReplicatedWriteRequest<Request>,
+            Response extends ReplicationResponse & ReplicatedWriteResponse
+        > extends TransportReplicationAction<Request, Request, Response> {
+
+    protected TransportWriteAction(Settings settings, String actionName, TransportService transportService,
+            ClusterService clusterService, IndicesService indicesService, ThreadPool threadPool, ShardStateAction shardStateAction,
+            ActionFilters actionFilters, IndexNameExpressionResolver indexNameExpressionResolver, Supplier<Request> request,
+            String executor) {
+        super(settings, actionName, transportService, clusterService, indicesService, threadPool, shardStateAction, actionFilters,
+                indexNameExpressionResolver, request, request, executor);
+    }
+
+    /**
+     * Called on the primary with a reference to the {@linkplain IndexShard} to modify.
+     */
+    protected abstract WriteResult<Response> onPrimaryShard(Request request, IndexShard indexShard) throws Exception;
+
+    /**
+     * Called once per replica with a reference to the {@linkplain IndexShard} to modify.
+     *
+     * @return the translog location of the {@linkplain IndexShard} after the write was completed or null if no write occurred
+     */
+    protected abstract Translog.Location onReplicaShard(Request request, IndexShard indexShard);
+
+    @Override
+    protected final WritePrimaryResult shardOperationOnPrimary(Request request) throws Exception {
+        final ShardId shardId = request.shardId();
+        IndexService indexService = indicesService.indexServiceSafe(shardId.getIndex());
+        IndexShard indexShard = indexService.getShard(shardId.id());
+        WriteResult<Response> result = onPrimaryShard(request, indexShard);
+        return new WritePrimaryResult(request, result.getResponse(), result.getLocation(), indexShard);
+    }
+
+    @Override
+    protected final WriteReplicaResult shardOperationOnReplica(Request request) {
+        final ShardId shardId = request.shardId();
+        IndexService indexService = indicesService.indexServiceSafe(shardId.getIndex());
+        IndexShard indexShard = indexService.getShard(shardId.id());
+        Translog.Location location = onReplicaShard(request, indexShard);
+        return new WriteReplicaResult(indexShard, request, location);
+    }
+
+    /**
+     * Simple result from a write action. Write actions have static method to return these so they can integrate with bulk.
+     */
+    public static class WriteResult<Response extends ReplicationResponse> {
+        private final Response response;
+        private final Translog.Location location;
+
+        public WriteResult(Response response, @Nullable Location location) {
+            this.response = response;
+            this.location = location;
+            // Set the ShardInfo to 0 so we can safely send it to the replicas
```

Comment:
I think it's there in original code because it is only used to populate the individual item responses in a shard bulk request. Those do not set the shardInfo on each item but rather on the request as a whole. This seems wasteful but I'm OK with doing it in another PR. Note though that I do think it's important to verify that in this version of the code it does the same.

Here is the comment from the old code (it's slightly different):

```
// this sets total, pending and failed to 0 and this is ok, because we will embed this into the replica
// request and not use it
```

I think they mean replication request not replica request...